lib/src/eol.rs

```diff
@@ -46,26 +80,82 @@ impl TargetEolStrategy {
     /// The limit is to probe whether the file is binary is 8KB.
     const PROBE_LIMIT: u64 = 8 << 10;
 
-    pub(crate) async fn convert_eol_for_snapshot<'a>(
+    pub(crate) async fn convert_eol_for_snapshot<'a, R, E>(
         &self,
         mut contents: impl AsyncRead + Send + Unpin + 'a,
-    ) -> Result<Box<dyn AsyncRead + Send + Unpin + 'a>, std::io::Error> {
+        read_old_contents: impl AsyncFnOnce() -> Result<Merge<Option<R>>, E>,
```

Comment:
nit: maybe you might expect, and I know you would disagree, but I would let the caller check contents of the old/new files because this function is getting complicated.